server/schedule/operator/step.go

```diff
@@ -437,3 +449,323 @@ func (al AddLightLearner) Influence(opInfluence OpInfluence, region *core.Region
 	to.RegionSize += region.GetApproximateSize()
 	to.RegionCount++
 }
+
+// DemoteFollower is an OpStep that demotes a region voter peer (not a leader) to learner.
+type DemoteFollower struct {
+	ToStore, PeerID uint64
+}
+
+func (df DemoteFollower) String() string {
+	return fmt.Sprintf("demote follower peer %v on store %v to learner", df.PeerID, df.ToStore)
+}
+
+// ConfVerChanged returns how much the conf version will increase due to this step.
+func (df DemoteFollower) ConfVerChanged(region *core.RegionInfo) uint64 {
+	peer := region.GetStoreLearner(df.ToStore)
+	return b2u64(peer.GetId() == df.PeerID)
+}
+
+// IsFinish checks if current step is finished.
+func (df DemoteFollower) IsFinish(region *core.RegionInfo) bool {
+	if peer := region.GetStoreLearner(df.ToStore); peer != nil {
+		if peer.Id != df.PeerID {
+			log.Warn("obtain unexpected peer", zap.String("expect", df.String()), zap.Uint64("obtain-learner", peer.Id))
+		}
+		return peer.Id == df.PeerID
+	}
+	return false
+}
+
+// CheckSafety checks if the step meets the safety properties.
+func (df DemoteFollower) CheckSafety(region *core.RegionInfo) error {
+	peer := region.GetStorePeer(df.ToStore)
+	if peer == nil || peer.Id != df.PeerID {
+		return errors.New("peer does not exist")
+	}
+	if peer.Id == region.GetLeader().GetId() {
+		return errors.New("peer is a leader, can not demote")
+	}
+	return nil
+}
+
+// Influence calculates the store difference that current step makes.
+func (df DemoteFollower) Influence(opInfluence OpInfluence, region *core.RegionInfo) {}
```

Comment:
Does it mean no influence here or will finish it later?

Reply:
yes，just like `PromoteLearner`